### PR TITLE
Add authentication to Stork

### PR DESCRIPTION
**Pull request by pault84**
Signed-off-by: Paul <paul@portworx.com>

**What type of PR is this?**
feature

**What this PR does / why we need it**:
This PR adds Authentication support to STORK for the OSD SDK

**Does this PR change a user-facing CRD or CLI?**:
no

**Is a release note needed?**:
yes

```release-note
Authentication support added, will read user-context from CRD annotations
```

**Does this change need to be cherry-picked to a release branch?**:
yes
2.1

### CODE REVIEW

File: drivers/volume/portworx/portworx.go [hunk 1]
```diff
@@ -158,22 +167,28 @@ func (p *portworx) Init(_ interface{}) error {
 	}
 
 	logrus.Infof("Using %v as endpoint for portworx volume driver", endpoint)
-	endpointPort := fmt.Sprintf("http://%v:%v", endpoint, apiPort)
+	tls := os.Getenv("stork-encryption-enabled")
```
Comment (disrani-px) [inline drivers/volume/portworx/portworx.go]: Make the env variable a const. Also, use upper case for env variable name

Comment (disrani-px) [reply drivers/volume/portworx/portworx.go]: Also, this isn't a stork config, but a PX config, so use something like `PX_ENABLE_TLS`

File: drivers/volume/portworx/portworx.go [hunk 2]
```diff
@@ -216,7 +231,12 @@ func (p *portworx) startNodeCache() error {
 }
 
 func (p *portworx) InspectVolume(volumeID string) (*storkvolume.Info, error) {
-	vols, err := p.volDriver.Inspect([]string{volumeID})
+	var err error
```
Comment (disrani-px) [inline drivers/volume/portworx/portworx.go]: Don't need this line

File: drivers/volume/portworx/portworx.go [hunk 3]
```diff
@@ -452,12 +472,95 @@ func (p *portworx) getSnapshotName(tags *map[string]string) string {
 	return "snapshot-" + (*tags)[snapshotter.CloudSnapshotCreatedForVolumeSnapshotUIDTag]
 }
 
+func (p *portworx) getUserVolDriver(annotations map[string]string) (volume.VolumeDriver, error) {
+	var err error
```
Comment (disrani-px) [inline drivers/volume/portworx/portworx.go]: Don't need to declare err and token

File: drivers/volume/portworx/portworx.go [hunk 4]
```diff
@@ -452,12 +472,95 @@ func (p *portworx) getSnapshotName(tags *map[string]string) string {
...
+	clnt, err := getClient("", endpointPort)
+	if err != nil {
+		return nil, err
+	}
+	return volumeclient.VolumeDriver(clnt), nil
+
+}
+
+func (p *portworx) getAdminVolDriver() (volume.VolumeDriver, error) {
+	secret := os.Getenv("stork-shared-secret")
```
Comment (disrani-px) [inline drivers/volume/portworx/portworx.go]: Use PX_SHARED_SECRET from a const

File: drivers/volume/portworx/portworx.go [hunk 5]
```diff
@@ -452,12 +472,95 @@ func (p *portworx) getSnapshotName(tags *map[string]string) string {
...
+	if len(secret) != 0 {
+		claims := &auth.Claims{
+			Issuer: "stork.openstorage.io",
+			Name:   "Stork",
+			Email:  "stork@openstorage.io",
```
Comment (disrani-px) [inline drivers/volume/portworx/portworx.go]: Does that email even exist?

File: drivers/volume/portworx/portworx.go [hunk 6]
```diff
@@ -452,12 +472,95 @@ func (p *portworx) getSnapshotName(tags *map[string]string) string {
...
+			Roles:  []string{"system.stork"},
+		}
+
+		signature := &auth.Signature{
+			Type: jwt.SigningMethodHS256,
+			Key:  []byte(secret),
+		}
+
+		options := &auth.Options{
+			Expiration: time.Now().Add(time.Second * 30).Unix(),
```
Comment (disrani-px) [inline drivers/volume/portworx/portworx.go]: Use a constant for the expiration time

File: drivers/volume/portworx/portworx.go [hunk 7]
```diff
@@ -452,12 +472,95 @@ func (p *portworx) getSnapshotName(tags *map[string]string) string {
...
+	if err != nil {
+		return nil, err
+	}
+	return volumeclient.VolumeDriver(clnt), nil
+}
+
+func getClient(endpointPort string, token string) (*apiclient.Client, error) {
+	if len(token) != 0 {
+		clnt, err := volumeclient.NewAuthDriverClient(endpointPort, "pxd", "", token, "", "stork")
+		if err != nil {
```
Comment (disrani-px) [inline drivers/volume/portworx/portworx.go]: Just `return volumeclient.NewAuthDriverClient(...)`

File: drivers/volume/portworx/portworx.go [hunk 8]
```diff
@@ -452,12 +472,95 @@ func (p *portworx) getSnapshotName(tags *map[string]string) string {
...
+			return nil, err
+		}
+		return clnt, nil
+	}
+	clnt, err := volumeclient.NewDriverClient(endpointPort, "pxd", "", "stork")
```
Comment (disrani-px) [inline drivers/volume/portworx/portworx.go]: Just `return volumeclient.NewDriverClient()`

File: drivers/volume/portworx/portworx.go [hunk 9]
```diff
@@ -452,12 +472,95 @@ func (p *portworx) getSnapshotName(tags *map[string]string) string {
...
+	if err != nil {
+		return nil, err
+	}
+	return clnt, nil
+
+}
+
+// User Context
```
Comment (disrani-px) [inline drivers/volume/portworx/portworx.go]: Remove these comments just mentioning the context

File: drivers/volume/portworx/portworx.go [hunk 10]
```diff
@@ -874,8 +994,18 @@ func (p *portworx) waitForCloudSnapCompletion(
 
 // getCloudSnapStatus fetches the cloudsnapshot status for given op and cloudsnap task ID
 func (p *portworx) getCloudSnapStatus(op api.CloudBackupOpType, taskID string) cloudSnapStatus {
-	response, err := p.volDriver.CloudBackupStatus(&api.CloudBackupStatusRequest{
-		Name: taskID,
+	var err error
+	volDriver, err := p.getAdminVolDriver()
```
Comment (disrani-px) [inline drivers/volume/portworx/portworx.go]: I think this should be user context

File: drivers/volume/portworx/portworx.go [hunk 11]
```diff
@@ -1569,6 +1729,12 @@ func (p *portworx) generateStatusReponseFromTaskIDs(
 
 // revertPXCloudSnaps deletes all cloudsnaps with given IDs
 func (p *portworx) revertPXCloudSnaps(cloudSnapIDs []string, credID string) {
+	var err error
+	volDriver, err := p.getAdminVolDriver()
+	if err != nil {
+		return
```
Comment (disrani-px) [inline drivers/volume/portworx/portworx.go]: Need to log error here

File: drivers/volume/portworx/portworx.go (hunk 10 above)
Comment (pault84) [reply drivers/volume/portworx/portworx.go]: We don't have access to a CRD or other metadata here to get the user context.

File: drivers/volume/portworx/portworx.go (hunk 5 above)
Comment (pault84) [reply drivers/volume/portworx/portworx.go]: removed it.

File: Gopkg.toml [hunk 12]
```diff
@@ -40,7 +40,7 @@ required = [
 
 [[constraint]]
   name = "github.com/libopenstorage/openstorage"
-  branch = "release-4.0"
+  branch = "release-6.0"
```
Comment (disrani-px) [inline Gopkg.toml]: Split the vendor changes into a separate commit

File: drivers/volume/portworx/portworx.go (hunk 10 above)
Comment (disrani-px) [reply drivers/volume/portworx/portworx.go]: Yeah pass it down from the caller of the API

**Comment by disrani-px** [comment]
Since this is updating all the APIs please run the integration tests with auth enabled and disabled with Portworx

File: drivers/volume/portworx/portworx.go [hunk 13]
```diff
@@ -99,8 +105,13 @@ const (
 
 	validateSnapshotTimeout       = 5 * time.Minute
 	validateSnapshotRetryInterval = 10 * time.Second
+
+	pxEnableTLS    = "PX_ENABLE_TLS"
+	pxSharedSecret = "PX_SHARED_SECRET"
 )
 
+var endpointPort string
```
Comment (disrani-px) [inline drivers/volume/portworx/portworx.go]: Move this to the `portworx` struct. Make all the methods accessing this variable members of the struct too

**Comment by disrani-px** [comment]
Closing as this has been fixed with #324
